PyTorch C++ API grammatically up-to-date

belphegor-s opened the pull request:
Grammatically up-to-date **PyTorch C++ API documentation** 

_Screenshot Provided\:_ 
[Screenshot 01/10/2020](https://ibb.co/xF0Jynq)

belphegor-s committed:
PyTorch C++ API grammatically up-to-date

Commit:
Revert "Revert D24024606: [FX] Shape propagation example" (#45637)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45637

This reverts commit 869b05648def7a3b01685da94d4ee36f671d5dd6.

Test Plan: Imported from OSS

Reviewed By: gmagogsfm

Differential Revision: D24037870

Pulled By: jamesr66a

fbshipit-source-id: 851beb42fe72383108ceeff1fe97f388d9ad059e

Commit:
Revert D23913105: [quant][graphmode][fx] Merge all quantization mode

Test Plan: revert-hammer

Differential Revision:
D23913105 (https://github.com/pytorch/pytorch/commit/ffcb0989e76798ddb893b9e156ae1113d2498bb5)

Original commit changeset: 4e335286d6de

fbshipit-source-id: 5765b4e8ec917423f1745f73a9f3f235fc53423d

Commit:
[AutoAccept][Codemod][FBSourceGoogleJavaFormatLinter] Daily `arc lint --take GOOGLEJAVAFORMAT`

Reviewed By: zertosh

Differential Revision: D24044052

fbshipit-source-id: 50ac5b7480ed65af94617bf8b014252ea7b27c4f

Commit:
[AutoAccept][Codemod][FBSourceClangFormatLinter] Daily `arc lint --take CLANGFORMAT`

Reviewed By: zertosh

Differential Revision: D24044108

fbshipit-source-id: 6dfe2f1201304fa58e42472e3f53c72cbb63d7d2

Commit:
Use MTA for amp grad unscaling, enforce op math type in MTA functors, and allow op lambdas (#44778)

Summary:
Amp gradient unscaling is a great use case for multi tensor apply (in fact it's the first case I wrote it for).  This PR adds an MTA unscale+infcheck functor.  Really excited to have it for `torch.cuda.amp`. izdeby your interface was clean and straightforward to use, great work!

Labeled as bc-breaking because the native_functions.yaml exposure of unscale+infcheck changes from [`_amp_non_finite_check_and_unscale_` to `_amp_foreach_non_finite_check_and_unscale_`]( https://github.com/pytorch/pytorch/pull/44778/files#diff-f1e4b2c15de770d978d0eb77b53a4077L6289-L6293).

The PR also modifies Unary/Binary/Pointwise Functors to
- do ops' internal math in FP32 for FP16 or bfloat16 inputs, which improves precision ([and throughput, on some architectures!](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#arithmetic-instructions)) and has no downside for the ops we care about.
- accept an instantiated op functor rather than an op functor template (`template<class> class Op`).  This allows calling code to pass lambdas.

Open question:  As written now, the PR has MTA Functors take care of pre- and post-casting FP16/bfloat16 inputs to FP32 before running the ops.  However, alternatively, the pre- and post-math casting could be deferred/written into the ops themselves, which gives them a bit more control.  I can easily rewrite it that way if you prefer.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/44778

Reviewed By: gchanan

Differential Revision: D23944102

Pulled By: izdeby

fbshipit-source-id: 22b25ccad5f69b413c77afe8733fa9cacc8e766d

izdeby committed:
Update torchvision version to current latest master (#45342)

Summary:
Updating torchvision version to the current latest master.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45342

Reviewed By: seemethere

Differential Revision: D23933572

Pulled By: izdeby

fbshipit-source-id: c374156eb608e882a1e2107143e39f03b7399081

ezyang committed:
Switch all Sequences in tools.codegen.model to Tuple (#45127)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45127

I thought I was being clever by using Sequence, which doesn't commit to
List or Tuple, but forces read-onlyness in the type system.  However,
there is runtime implication to using List or Tuple: Lists can't be
hashed, but Tuples can be!  This is important because I shortly want
to group by FunctionSchema, and to do this I need FunctionSchema to
be hashable.  Switch everything to Tuple for true immutability.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Reviewed By: gchanan

Differential Revision: D23872527

Pulled By: ezyang

fbshipit-source-id: 5c8fae1c50a5ae47b4167543646d94ddcafff8c3

ezyang committed:
Add NativeFunction.signature and kind. (#45131)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45131

These make it easier to group native functions together and determine
what kind of native function it is (inplace/out/functional).  Currently
they are not used but they may be useful for tools.autograd porters.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Reviewed By: zhangguanheng66

Differential Revision: D23872526

Pulled By: ezyang

fbshipit-source-id: 1d6e429ab9a1f0fdb764be4228c5bca4dce8f24e

IvanYashchuk committed:
Added support for complex torch.symeig (#45121)

Summary:
This PR adds support for complex-valued input for `torch.symeig`.

TODO:
- [ ] complex cuda tests raise `RuntimeError: _th_bmm_out not supported on CUDAType for ComplexFloat`
Update: Added xfailing tests for complex dtypes on CUDA. Once support for complex `bmm` is added these tests will work.

Fixes https://github.com/pytorch/pytorch/issues/45061.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45121

Reviewed By: mrshenli

Differential Revision: D24049649

Pulled By: anjali411

fbshipit-source-id: 2cd11f0e47d37c6ad96ec786762f2da57f25dac5

zou3519 committed:
[docs] Fix back quote rendering in loss modules docs (#45662)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/42855. Previously, back quotes weren't rendering correctly in
equations. This is because we were quoting things like `'mean'`. In
order to backquote properly in latex in text-mode, the back-quote needs
to be written as a back-tick.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45662

Test Plan:
- built docs locally and viewed the changes.

For NLLLoss (which is not the original module mentioned in the issue, but it has the same problem), we can see how the back quotes now render properly:

![image](https://user-images.githubusercontent.com/5652049/94819862-c5676a00-03cd-11eb-9e92-01380ee52bd6.png)

Reviewed By: glaringlee

Differential Revision: D24049880

Pulled By: zou3519

fbshipit-source-id: 61a1257994144549eb8f29f19d639aea962dfec0

zou3519 committed:
[docs] Fix hyperlinks for nn.CrossEntropyLoss (#45660)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/45460. This PR makes it so that LogSoftmax and NLLLoss are correctly linked from the nn.CrossEntropyLoss documentation.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45660

Test Plan:
- built and viewed docs locally

![image](https://user-images.githubusercontent.com/5652049/94816513-ee85fb80-03c9-11eb-8289-56642c133e11.png)

Reviewed By: glaringlee

Differential Revision: D24049009

Pulled By: zou3519

fbshipit-source-id: 3bd0660acb8575d753cefd2d0f1e523ca58a25b6

Krovatkin committed:
Disable tcuda_fuser tests in Profiling Mode (#45638)

Summary:
Disable tcuda_fuser tests in Profiling Mode to address flakey tests until fuser switches to the new approach.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45638

Reviewed By: mrshenli

Differential Revision: D24057230

Pulled By: Krovatkin

fbshipit-source-id: 8f7a47610d9b7da6ad3057208057a5a596e1bffa

scottxu0730 committed:
Find single partition (#45429)

Summary:
WIP: This PR is working in progress for the partition of fx graph module. _class partitioner_ generates partitions for the graph module. _class partition_ is a partition node in the partitions.
_Partitioner()_ : create a partitioner
_partition_graph(self, fx_module: GraphModule, devices: List[str]) -> None_:
use fx graph module and devices as the input and create partition_ids for each node inside the graph module

_dump_partition_DAG(self) -> None_:
print out the information about each partition, including its id, its backend type (what type of device this partition uses), all the nodes included in this partition,  its parent partitions, children partitions, input nodes, and output nodes.

So far, only a single partition is considered, which means there is only one device with unlimited memory.
A test unit call _test_find_single_partition()_ is added to test if all nodes in the graph are marked for the only partition.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45429

Reviewed By: izdeby

Differential Revision: D24026268

Pulled By: scottxu0730

fbshipit-source-id: 119d506f33049a59b54ad993670f4ba5d8e15b0b

Commit:
[RFC][caffe2] TaskGroup.__repr__ shouldn't have side effects

Summary: `__repr__` calling self.tasks() ends up marking the instance as "used", which doesn't seem appropriate. I was debugging a value being passed around and then ran into `Cannot add Task to an already used TaskGroup.` because the value had been logged once.

Test Plan:
Added a unit test -- didn't see a clean public method to test it, but I'm happy to add one if that makes sense.

Will wait for sandcastle to trigger everything else; I'm not at all familiar with this code so any other recommendations would be great!

Reviewed By: cryptopic

Differential Revision: D23541198

fbshipit-source-id: 5d1ec674a1ddaedf113140133b90e0da6afa7270

Commit:
[quant] torch.mean add path for unsupported QNNPACK modes (#45533)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45533

Test Plan: Imported from OSS

Reviewed By: jerryzh168

Differential Revision: D24030446

Pulled By: z-a-f

fbshipit-source-id: a392402ef701c5e45e244ac440bc151ef942cccd

seemethere committed:
setup: Only include dataclasses for py < 3.8 (#45611)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45611

dataclasses was made a standard library item in 3.8

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Test Plan: Imported from OSS

Reviewed By: walterddr

Differential Revision: D24031740

Pulled By: seemethere

fbshipit-source-id: 15bdf1fe0d8de9b8ba7912e4a651f06b18d516ee

Commit:
Fix BUILD_CAFFE2 if FBGEMM and NNPACK are not built (#45610)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45610

Also add to the usual documentation places that this option exists.

Test Plan: Imported from OSS

Reviewed By: gmagogsfm

Differential Revision: D24058199

Pulled By: suo

fbshipit-source-id: 81574fbd042f47587e2c7820c726fac0f68af2a7

jerryzh168 committed:
[reland][quant][graphmode][fx] Merge all quantization mode (#45292) (#45672)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45672

This PR merges all quantization mode and will only expose the following top level functions:
```
prepare_fx
prepare_qat_fx
convert_fx
```

Test Plan:
Imported from OSS

Imported from OSS

Reviewed By: z-a-f

Differential Revision: D24053439

fbshipit-source-id: 03d545e26a36bc22a73349061b751eeb35171e64

Commit:
Use CAFFE2_USE_MSVC_STATIC_RUNTIME to determine when to avoid waiting for global destructors on Windows (#43532)

Summary:
We are trying to build libtorch statically (BUILD_SHARED_LIBS=OFF) then link it into a DLL. Our setup hits the infinite loop mentioned [here](https://github.com/pytorch/pytorch/blob/54c05fa34e1fbbb5096746a8ae92e27a08116de4/torch/csrc/autograd/engine.cpp#L228) because we build with `BUILD_SHARED_LIBS=OFF` but still link it all into a DLL at the end of the day.

This PR fixes the issue by changing the condition to guard on which windows runtime the build links against using the `CAFFE2_USE_MSVC_STATIC_RUNTIME` flag. `CAFFE2_USE_MSVC_STATIC_RUNTIME` defaults to ON when `BUILD_SHARED_LIBS=OFF`, so backwards compatibility is maintained.

I'm not entirely confident I understand the subtleties of the windows runtime versus linking setup, but this setup works for us and should not affect the existing builds.

Fixes https://github.com/pytorch/pytorch/issues/44470

Pull Request resolved: https://github.com/pytorch/pytorch/pull/43532

Reviewed By: mrshenli

Differential Revision: D24053767

Pulled By: albanD

fbshipit-source-id: 1127fefe5104d302a4fc083106d4e9f48e50add8

xta0 committed:
[iOS][CI] Upgrade xcode version to 12.0 (#45677)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45677

Test Plan: Imported from OSS

Reviewed By: husthyc

Differential Revision: D24065647

Pulled By: xta0

fbshipit-source-id: f2535b1d93e58cf79e7075bf56b0613a3ded16eb

xta0 committed:
[iOS][CI] Update the dev cert (#45651)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45651

### Summary

1. Update the iOS developer certificates. The new expiration date is 10/01/2021.
2. Restore the iOS arm64 jobs and the nightly.

### Test Plan

The following CI jobs succeed

- ci/circleci: pytorch_ios_11_2_1_arm64_build
- ci/circleci: pytorch_ios_11_2_1_arm64_custom_build
- ci/circleci: pytorch_ios_11_2_1_x86_64_build

Test Plan: Imported from OSS

Reviewed By: husthyc

Differential Revision: D24065648

Pulled By: xta0

fbshipit-source-id: 758f41de8296fdfbd3cfad87e9445c2acafd5f94

Krovatkin committed:
add an indexer to SymbolicShape (#45450)

Summary:
A convenience indexer into `SymbolicShape`s

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45450

Reviewed By: ZolotukhinM

Differential Revision: D23971758

Pulled By: Krovatkin

fbshipit-source-id: 1f18c5f89f579072f6bf467809ea9471bf42bc2d

glaringlee commented on docs/cpp/source/index.rst:
```diff
@@ -1,26 +1,26 @@
-Together, these building blocks form a research and
-production ready C++ library for tensor computation and dynamic neural
+Combining, these building blocks form a research and
+production ready C++ library for the tensor computation and dynamic neural
```
please remove `the` here since we have two items mentioned here. (with `and`)

belphegor-s replied:
i have mad the changes plz confirm

glaringlee commented on docs/cpp/source/index.rst:
```diff
@@ -1,26 +1,26 @@
-  make major breaking changes to the backend in order to improve the API,
-  or in service of providing the Python interface to PyTorch, which is our
+  make major breaking changes to the backend in order to improvise the API,
+  or in service of providing the Python interface to PyTorch, which is the our
```
Please remove `the` here, not necessary

belphegor-s replied:
made the changes plz confirm

glaringlee commented on docs/cpp/source/index.rst:
```diff
@@ -1,26 +1,26 @@
   At the moment, the C++ API should be considered "beta" stability; we may
-  make major breaking changes to the backend in order to improve the API,
-  or in service of providing the Python interface to PyTorch, which is our
+  make major breaking changes to the backend in order to improvise the API,
```
I think `improve` is more suitable here than `improvise`

glaringlee reviewed (changes requested):
Please see my comments

Krovatkin committed:
win job for the legacy executor (#45612)

Summary:
Adds a CUDA job on Windows for the jit legacy executor

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45612

Reviewed By: mrshenli

Differential Revision: D24042196

Pulled By: Krovatkin

fbshipit-source-id: 35c79c53ed569d221e79376c108bc864900ef49e

Krovatkin committed:
remove skip annotations since we already disabled the tests wholesale (#45698)

Summary:
Remove skip annotations since we already disabled the tests wholesale

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45698

Reviewed By: mrshenli

Differential Revision: D24064547

Pulled By: Krovatkin

fbshipit-source-id: 0d154135de0c0550d6874bea3c2d42d5f4d71cb4

Commit:
Fix the bucketization wrong doc for right argument (#45684)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45684

Test Plan: Imported from OSS

Reviewed By: zou3519

Differential Revision: D24057996

Pulled By: glaringlee

fbshipit-source-id: 3db1c24f3cae9747effa4b1f3c5c3baf6888c9a1

bdhirsh committed:
migrate cuda implementation of take() from TH to ATen (#45430)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45430

Test Plan: Imported from OSS

Reviewed By: mrshenli

Differential Revision: D24037297

Pulled By: bdhirsh

fbshipit-source-id: 7c5f2c08e895fb0c25eec1d68c7455e4f2b1c64e

bdhirsh committed:
migrating the take() fn from TH to ATen (#45283)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45283

Test Plan: Imported from OSS

Reviewed By: mrshenli

Differential Revision: D24037298

Pulled By: bdhirsh

fbshipit-source-id: 088ce39e55ee8b5a79fa501395fa9eec08d1d396

Commit:
example of splitting up an FX graph into smaller subgraphs with own submodules (#45404)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45404

Test Plan: Imported from OSS

Reviewed By: jamesr66a

Differential Revision: D23956147

Pulled By: Lilyjjo

fbshipit-source-id: a35e33a0b9f1ed5f3fb6e5cd146f66c29bf3d518

jeffdaily committed:
passing all arguments to sccache wrapper script should be quoted as "$@" (#45582)

Summary:
This fixes MIOpen runtime compilation since it passes quoted arguments to the clang compiler.  This change also makes the sccache wrapper scripts consistent with the nvcc wrapper.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45582

Reviewed By: seemethere, izdeby

Differential Revision: D24034477

Pulled By: malfet

fbshipit-source-id: 1964bac1e693b238e8efe9c046a39be64571e9df

Commit:
[docs] Update docs for NegativeBinomial (#45693)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45693

**Summary**
This commit updates the docstring for
`torch.distributions.NegativeBinomial` to better match actual behaviour.
In particular, the parameter currently documented as probability of
success is actually probability of failure.

**Test Plan**
1) Ran the code from the issue to make sure this is still an issue (it
is)
2) `make html` and viewed the docs in a browser.

*Before*
<img width="879" alt="Captura de Pantalla 2020-10-01 a la(s) 1 35 28 p  m" src="https://user-images.githubusercontent.com/4392003/94864456-db3a5680-03f0-11eb-977e-3bab0fb9c206.png">

*After*
<img width="877" alt="Captura de Pantalla 2020-10-01 a la(s) 2 12 24 p  m" src="https://user-images.githubusercontent.com/4392003/94864478-e42b2800-03f0-11eb-965a-51493ca27c80.png">

**Fixes**
This commit closes #42449.

Test Plan: Imported from OSS

Reviewed By: robieta

Differential Revision: D24071048

Pulled By: SplitInfinity

fbshipit-source-id: d345b4de721475dbe26233e368af62eb57a47970

Krovatkin committed:
Set the default bailout depth to 20 (#45710)

Summary:
This modifies the default bailout depth to 20 which gives us a reasonable performance in benchmarks we considered (fastrnns, maskrcnn, hub/benchmark, etc)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45710

Reviewed By: robieta

Differential Revision: D24071861

Pulled By: Krovatkin

fbshipit-source-id: 472aacc136f37297b21f577750c1d60683a6c81e

supriyar committed:
[quant] PerChannelFloatQParams support for quint4x2 dtype (#45594)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45594

Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Imported from OSS

Reviewed By: raghuramank100

Differential Revision: D24025595

fbshipit-source-id: dd9d0557de585dd4aaf5f138959c3523a29fb759

supriyar committed:
[quant] creating quint4x2 dtype for quantized tensors (#44678)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/44678

This is a prototype PR that introduces 4 bit qtensors. The new dtype added for this is c10::quint4x2
The underlying storage for this is still uint8_t, so we pack 2 4-bit values in a byte while quantizing it.

This change uses most of the existing scaffolding for qtensor storage. We allocate storage
based on the dtype before creating a new qtensor.

It also adds a dispatch mechanism for this dtype so we can use this to get the bitwidth, qmin and qmax info
while quantizing and packing the qtensor (when we add 2-bit qtensor)

Kernels that use this dtype should be aware of the packing format.

Test Plan:
Locally tested
```
x = torch.ones((100, 100), dtype=torch.float)
qx_8bit = torch.quantize_per_tensor(x, scale=1.0, zero_point=2, dtype=torch.quint8)
qx = torch.quantize_per_tensor(x, scale=1.0, zero_point=2, dtype=torch.quint4x2)

torch.save(x, "temp.p")
print('Size float (B):', os.path.getsize("temp.p"))
os.remove('temp.p')

torch.save(qx_8bit, "temp.p")
print('Size quantized 8bit(B):', os.path.getsize("temp.p"))
os.remove('temp.p')

torch.save(qx, "temp.p")
print('Size quantized 4bit(B):', os.path.getsize("temp.p"))
os.remove('temp.p')
```

Size float (B): 40760
Size quantized 8bit(B): 10808
Size quantized 4bit(B): 5816

Imported from OSS

Reviewed By: raghuramank100

Differential Revision: D23993134

fbshipit-source-id: 073bf262f9680416150ba78ed2d932032275946d

belphegor-s committed:
updated index.rst

Commit:
Use addmm directly for 1x1 convolution (#45557)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/45274
Based on https://github.com/pytorch/pytorch/issues/44041, sets intermediate for backward computation (otherwise, backward tests are failing).

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45557

Reviewed By: izdeby

Differential Revision: D24030655

Pulled By: ngimel

fbshipit-source-id: 368fe9440668dffc004879f8b1d2dd3787d915c9

smessmer committed:
Add faithful C++ API (#44087)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/44087

Each op taking a TensorOptions argument now has an additional overload in the C++ frontend where it takes scattered ScalarType, Layout, Device, bool instead of one TensorOptions argument.

If it is a c10-full op, then the scattered version calls into the dispatcher and the gathered version is a proxy calling into the scattered version.
If it is a non-c10-full op, then the gathered version calls into the dispatcher and the scattered version is a proxy calling into the gathered version.

This should minimize the amount of gathering and scattering needed.

This PR is also a prerequisite to remove the re-gathering of arguments that is currently happening in VariableKernel. Currently, VariableKernels gather arguments into a TensorOptions object
to call into the C++ API. In a PR stacked on top of this, VariableKernel will just directly call into the scattered C++ API introduced here and avoid the gathering step.
ghstack-source-id: 113355689

Test Plan:
waitforsandcastle

vs master: https://www.internalfb.com/intern/fblearner/details/216169815/

vs previous diff: https://www.internalfb.com/intern/fblearner/details/216169957/

Reviewed By: ezyang

Differential Revision: D23492188

fbshipit-source-id: 3e84c467545ad9371e98e09075a311bd18411c5a

smessmer committed:
VariableKernel calls into scattered C++ api (#44158)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/44158

Previously, the C++ API only supported calling ops with a gathered TensorOptions object. So even if the VariableKernel took scattered arguments,
it had to re-gather them to call into the C++ API. But a diff stacked below this one introduced a scattered API for the C++ frontend.

This reaps the benefits and makes sure that if the Variable kernel gets scattered arguments (i.e. it's a c10-full op), then it passes those on without regathering
ghstack-source-id: 113355690

Test Plan:
vs master: https://www.internalfb.com/intern/fblearner/details/216342597/

vs prev diff: https://www.internalfb.com/intern/fblearner/details/216342688/

Reviewed By: ezyang

Differential Revision: D23512538

fbshipit-source-id: 8ee6c1cc99443a2141db85072fd6dbc52b4d77fd

bdhirsh committed:
make broadcasting explanation clearer in matmul doc: #22763 (#45699)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45699

Test Plan: Imported from OSS

Reviewed By: mruberry

Differential Revision: D24065584

Pulled By: bdhirsh

fbshipit-source-id: 5e2cdd00ed18ad47d24d11751cfa5bee63853cc9

bdhirsh committed:
remove beta defaulting in smooth_l1_loss_backward. added to the bc whitelist (#45588)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45588

Test Plan: Imported from OSS

Reviewed By: mrshenli

Differential Revision: D24024312

Pulled By: bdhirsh

fbshipit-source-id: 7246e5da741fbc5641deecaf057ae9a6e44e8c34

bdhirsh committed:
some documentation and style fixes to smooth_l1_loss (#45587)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45587

Test Plan: Imported from OSS

Reviewed By: mrshenli

Differential Revision: D24024313

Pulled By: bdhirsh

fbshipit-source-id: c50efb2934d7b9d3b090e92678319cde42c0df45

glaringlee commented:
@belphegor-s 
Thanks a lot for correcting this. I will approve it, can you do a rebase? It seems your code is based on an old master branch which failed one of our github CI test.

malfet committed:
Test torch.svd using complex float and double numbers (#45572)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45572

Reviewed By: anjali411

Differential Revision: D24018160

Pulled By: malfet

fbshipit-source-id: 1b6103f5af94e9f74b73ed23aa02c0236b199b34

belphegor-s commented:
> @belphegor-s 
> Thanks a lot for correcting this. I will approve it, can you do a rebase? It seems your code is based on an old master branch which failed one of our github CI test.

    I don't have any idea how can I do that .. it would be a great do if you can suggest me how.

samestep committed:
Enable type check for torch.quantization.fake_quantize (#45701)

Summary:
Addresses part of https://github.com/pytorch/pytorch/issues/42969.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45701

Reviewed By: walterddr

Differential Revision: D24066672

Pulled By: samestep

fbshipit-source-id: 53bb5e7b4703738d3de86fa89fb0980f1d6251f3

mingzhe09088 committed:
Add a distributed package reviewer (#45744)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45744

Tag me as reviewer

Test Plan: na

Reviewed By: jiayisuse

Differential Revision: D23881569

fbshipit-source-id: 8452fa60fe3d017ae1f0da26c0ce476f2b9c170c

MarcioPorto committed:
Exposes SparseToDenseMask Caffe2 Operator (#45670)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45670

Reviewed By: esqu1

Differential Revision: D23868280

fbshipit-source-id: d6afa129c073fe611cb43a170025bc3c880a4bec

spaugh committed:
Add string versions of argument funcs in jit Node (#45464)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45464

Usage of Symbols to find arguments requires one to generate a nonsense symbol for inputs which don't already have one. The intention of symbols appears to be something of an internalized string, but the namespace component doesn't apply to an argument. In order to access the arguments by name without adding new symbols, versions of those functions with std::string input was added. These can be proved valid based on the existing codepath. Additionally, a hasNamedInput convenience function was added to remove the necessity of a try/catch block in user code.

The primary motivation is to be able to easily handle the variable number of arguments in glow, so that the arange op may be implemented.

Reviewed By: eellison

Differential Revision: D23972315

fbshipit-source-id: 3e0b41910cf07e916186f1506281fb221725a91b

pritamdamania committed:
Use correct signatures for METH_NOARGS. (#45528)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45528

As described in https://github.com/pytorch/pytorch/issues/45419,
resolving a bunch of cpython signature issues.

#Closes: https://github.com/pytorch/pytorch/issues/45419
ghstack-source-id: 113385726

Test Plan: sentinel

Reviewed By: albanD

Differential Revision: D24000626

fbshipit-source-id: d334596f1f0256063691aa044c8fb2face260817

seemethere committed:
Bump nightlies to 1.8.0 (#45696)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45696

Similar to https://github.com/pytorch/pytorch/pull/40519

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Test Plan: Imported from OSS

Reviewed By: samestep

Differential Revision: D24064381

Pulled By: seemethere

fbshipit-source-id: 1484b9c4fc5fa8cfa7be591a0a5d4b6e05968589

osalpekar committed:
[Docs] Adding Store API Docs (#45543)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45543

This PR adds documentation for the c10d Store to the public docs. Previously these docs were missing although we exposed a lightly-used (but potentially useful) Python API for our distributed key-value store.
ghstack-source-id: 113409195

Test Plan: Will verify screenshots by building the docs.

Reviewed By: pritamdamania87

Differential Revision: D24005598

fbshipit-source-id: 45c3600e7c3f220710e99a0483a9ce921d75d044

janeyx99 committed:
adding sharding option to run_test.py (#45583)

Summary:
Added a sharding option to run_test.py to enable users to run a subset of the many tests. The new `--shard` argument takes in two integer values, `x` and `y`, where the larger value would denote the number of shards and the smaller value would denote which shard to run.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45583

Reviewed By: malfet

Differential Revision: D24083469

Pulled By: janeyx99

fbshipit-source-id: 1777bd7822c95b3bf37079deff9381c6f8eaf4cc

Commit:
Add function signature for pixel_shuffle (#45661)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45661

Test Plan: Imported from OSS

Reviewed By: gmagogsfm

Differential Revision: D24078627

Pulled By: ansleyadelaide

fbshipit-source-id: 44917ff5932e4d0adcc18ce24ecfc0b5686818e3

Commit:
Change signature for torch.poisson (#45656)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45656

Test Plan: Imported from OSS

Reviewed By: gmagogsfm

Differential Revision: D24078609

Pulled By: ansleyadelaide

fbshipit-source-id: 97a95b08334ed0d710e032a267b940c2fc9f7f40

Commit:
type check for torch.quantization.observer (#45630)

Summary:
add type checker for observer

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45630

Reviewed By: malfet

Differential Revision: D24058304

Pulled By: walterddr

fbshipit-source-id: ac1c0f5ff0d34b0445bd1364653fc5c9d7571b05

glaringlee commented:
@pytorchbot rebase this please

xwang233 committed:
Fix cuDNN error message when it's Conv2d (#45729)

Summary:
Originally introduced in https://github.com/pytorch/pytorch/issues/45023. When I was doing test in the original PR, it was a Conv3d, so this problem was not discovered.

Arrays in `ConvolutionParams` have a fixed length of 3 or 5. This is because `max_dim` is set as a constexpr of 3, regardless of Conv2d or Conv3d. The current code will make some error message be weird. See below in the comments.

https://github.com/pytorch/pytorch/blob/9201c37d020007979e144693d86c8e8599e2fd8f/aten/src/ATen/native/cudnn/Conv.cpp#L212-L226

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45729

Reviewed By: mruberry

Differential Revision: D24081542

Pulled By: ngimel

fbshipit-source-id: 141f8946f4d0db63a723131775731272abeaa6ab

rohan-varma committed:
Enable NamedTuple data type to work with DDP (#44220)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/44220

Closes https://github.com/pytorch/pytorch/issues/44009
Currently if a dataloader returns objects created with a
collections.namedtuple, this will incorrectly be cast to a tuple. As a result, if we have data of these types, there can be runtime errors during the forward pass if the module is expecting a named tuple.

Fix this in
`scatter_gather.py` to resolve the issue reported in
https://github.com/pytorch/pytorch/issues/44009
ghstack-source-id: 113423287

Test Plan: CI

Reviewed By: colesbury

Differential Revision: D23536752

fbshipit-source-id: 3838e60162f29ebe424e83e474c4350ae838180b

malfet committed:
Revert D24018160: [pytorch][PR] Test torch.svd using complex float and double numbers

Test Plan: revert-hammer

Differential Revision:
D24018160 (https://github.com/pytorch/pytorch/commit/888f3c12e75c981cd5d3a458e75c1e421d6c514a)

Original commit changeset: 1b6103f5af94

fbshipit-source-id: 3040250db25995fc0d41fd0f497550dded43cad9

bertmaher committed:
[te] Get llvm codegen to compile with llvm9 and llvm-fb (#45726)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45726

FB has an old internal platform that uses some random llvm version
that looks sort of like llvm 7.  I've guarded that with the appropriate
LLVM_VERSION_PATCH.

I've also swapped out some of our uses of ThreadSafeModule/ThreadSafeContext
for the variants without ThreadSafe in the name.  As far as I can tell we
weren't using the bundled locks anyways, but I'm like 85% sure this is OK since
we compile under the Torch JIT lock anyways.

Test Plan: unit tests

Reviewed By: ZolotukhinM, asuhan

Differential Revision: D24072697

fbshipit-source-id: 7f56b9f3cbe5e6d54416acdf73876338df69ddb2

glaringlee commented:
> > @belphegor-s
> > Thanks a lot for correcting this. I will approve it, can you do a rebase? It seems your code is based on an old master branch which failed one of our github CI test.
> 
> ```
> I don't have any idea how can I do that .. it would be a great do if you can suggest me how. 
> ```

You can checkout your branch, and pull latest code from pytorch master branch and commit and push. Do the following:

git pull https://github.com/pytorch/pytorch --rebase
git commit --amend (you might need to enter ctrl+X after this commend)
git push

anjali411 committed:
Update backward definition for more operators and reenable tests in test_ops.py (#44444)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/44444

This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`, `angle`.
2. Re-enables the tests in `test_ops.py`.
3. Adds dispatch for complex dtypes for `tanh_backward`.
4. Re-enables commented tests in `common_methods_invocation.py`.

Test Plan: Imported from OSS

Reviewed By: glaringlee

Differential Revision: D24027761

Pulled By: anjali411

fbshipit-source-id: c1f707c2a039149a6e04bbde53ee120d9119d99a

MarcioPorto committed:
Torch Integration Test Formatting Changes (#45740)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45740

Reviewed By: esqu1

Differential Revision: D23869021

fbshipit-source-id: 5910d44f9475bd7a53dc0478b69b39572dc8666f

xianjiec committed:
[caffe2] Optimize Dedup version of RowWiseSparseAdagrad fused op by WarpReduce (#45649)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45649

Pull Request resolved: https://github.com/pytorch/pytorch/pull/44275

* This Diff applies WarpReduce optimization for dedup version of RowWiseSparseAdagrad fused op. Basically we can achieve ~1.33x performance improvement with this Diff.

* Port the way from D23948802 to find the num_dup
* fix the likely bug about fp16 in the dedup kernel

Reviewed By: jianyuh

Differential Revision: D23561994

fbshipit-source-id: 1a633fcdc924593063a67f9ce0d36eadb19a7efb

gmagogsfm committed:
Make sure each warnings.warn only executes once inside TorchScript. (#45382)

Summary:
* Add a pass at end of runCleanupPasses to annotate `aten::warn` so that each has its unique id
* Enhanced interpreter so that it tracks which `aten::warn` has been executed before and skip them
* Improved insertInstruction so that it correctly checks for overflow

Fixes https://github.com/pytorch/pytorch/issues/45108

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45382

Reviewed By: mrshenli

Differential Revision: D24060677

Pulled By: gmagogsfm

fbshipit-source-id: 9221bc55b9ce36b374bdf614da3fe47496b481c1

mrshenli committed:
Revert "Remove device maps from TensorPipe for v1.7 release (#45353)" (#45762)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45762

This reverts commit 5211fb97ac4c246151f1286c78d63e0e317a8a4a.

Test Plan: Imported from OSS

Reviewed By: colesbury

Differential Revision: D24088231

Pulled By: mrshenli

fbshipit-source-id: b6ee15ec5ae137ea127bdc2db8e1842764bc01d4

zasdfgbnm committed:
CUDA BFloat16 infrastructure (#44925)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/44925

Reviewed By: agolynski

Differential Revision: D23783910

Pulled By: ngimel

fbshipit-source-id: dacac2ad87d58056bdc68bfe0b7ab1de5c2af0d8

Commit:
[FX] Make output a non-special Node (#45599)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/45599

Test Plan: Imported from OSS

Reviewed By: zdevito

Differential Revision: D24027586

Pulled By: jamesr66a

fbshipit-source-id: 747c25e3c7668ca45f03bed0be71fd3c9af67286

suo committed:
Fix JIT tests when run locally in fbcode (#45776)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45776

Splitting out backend and custom class registration into their own library is
not currently implemented in fbcode, so detect that we are running tests in
fbcode and disable those tests.

Test Plan: buck test mode/no-gpu mode/dev caffe2/test:jit

Reviewed By: smessmer

Differential Revision: D24085871

fbshipit-source-id: 1fcc0547880bc4be59428e2810b6a7f6e50ef798

ezyang committed:
Revert D24027761: Update backward definition for more operators and reenable tests in test_ops.py

Test Plan: revert-hammer

Differential Revision:
D24027761 (https://github.com/pytorch/pytorch/commit/7d809f5d8e93cb6d332297ece071083845e30e26)

Original commit changeset: c1f707c2a039

fbshipit-source-id: 30750d2f08886036fb8b2cd0ae51c7732d3b7b19

belphegor-s commented:
> 
> 
> > > @belphegor-s
> > > Thanks a lot for correcting this. I will approve it, can you do a rebase? It seems your code is based on an old master branch which failed one of our github CI test.
> > 
> > 
> > ```
> > I don't have any idea how can I do that .. it would be a great do if you can suggest me how. 
> > ```
> 
> You can checkout your branch, and pull latest code from pytorch master branch and commit and push. Do the following:
> 
> git pull https://github.com/pytorch/pytorch --rebase
> git commit --amend (you might need to enter ctrl+X after this commend)
> git push

Thankyou for your take on this, I have done the rebase. Pls Confirm.

glaringlee commented:
@belphegor-s 

> > > > @belphegor-s
> > > > Thanks a lot for correcting this. I will approve it, can you do a rebase? It seems your code is based on an old master branch which failed one of our github CI test.
> > > 
> > > 
> > > ```
> > > I don't have any idea how can I do that .. it would be a great do if you can suggest me how. 
> > > ```
> > 
> > 
> > You can checkout your branch, and pull latest code from pytorch master branch and commit and push. Do the following:
> > git pull https://github.com/pytorch/pytorch --rebase
> > git commit --amend (you might need to enter ctrl+X after this commend)
> > git push
> 
> Thankyou for your take on this, I have done the rebase. Pls Confirm.

Hi, I did not see your new commit in the PR. The last commit is 22hr ago. Can you check? If you push your code to your branch, your pull request should be updated automatically.

glaringlee commented:
@belphegor-s 
I think I know what happened. 
You need to checkout your branch before doing the rebase, I went to your github pytorch repo, the master branch is rebased instead of pytorch_contri branch.

belphegor-s commented:
on it doing right now

belphegor-s commented:
Finally done please confirm..
![Capture](https://user-images.githubusercontent.com/65305602/94984026-650b3080-0565-11eb-9321-83b7140e939a.PNG)

glaringlee commented:
@belphegor-s 
I think something is wrong. Now your PR has 162 changed files.
Let's make it simple, can you just create a new PR and update the PR number here? I will then go ahead to approve it without any problem. Sorry about this inconvenience.

belphegor-s commented:
Ok

belphegor-s commented:
Hi , i have made another PR #45799 
[Click Here](https://github.com/pytorch/pytorch/pull/45799)

glaringlee commented:
> Hi , i have made another PR #45799
> [Click Here](https://github.com/pytorch/pytorch/pull/45799)

Hi, I think your branch has something wrong. Your new PR has 12 changed files, but you actually changed 1 file, so your branch has older changes. I think what you can do is backup your changed file, create a new branch from master and copy your changes in, and then create a new PR. Your pytorch-contri branch is messed up already I guess.

belphegor-s commented:
every PR i create there is that " you have 3 commits", i am totally mad. HELP ME OUT!

glaringlee commented:
@belphegor-s 
Sry, it was too late yesterday at my location. (EST)
If you don't have other thing in your pytorch repo, you can save your modified file, delete your pytorch repo from github and re-fork one, this will be fresh and no problem at all.

belphegor-s commented:
HEY I HAVE CREATED ANOTHER PR , as mentioned i have deleted the whole repo and then forked the new one , made the changes through GITHUB itself and commited them.... Please take a look if you can review that..
__MY PR \:__
[Click ME](https://github.com/pytorch/pytorch/pull/45801)
